### ISSUE TITLE
[Distributed] Not properly saving record?

### ISSUE THREAD

**Issue by Dacnomanie** [orientechnologies/orientdb#6433]
When I run the following for the second time (the second time is when it throws the error, this first still does not save to the other server) this error is thrown:

`com.orientechnologies.orient.core.exception.ODatabaseException: Error during saving of record with rid #27:0
	DB name="Carnoot"
	at com.orientechnologies.orient.core.tx.OTransactionNoTx.saveRecord(OTransactionNoTx.java:205)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:2541)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:101)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1796)
	at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1787)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLUpdate.result(OCommandExecutorSQLUpdate.java:353)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLResultsetAbstract.pushResult(OCommandExecutorSQLResultsetAbstract.java:278)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.addResult(OCommandExecutorSQLSelect.java:735)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.handleResult(OCommandExecutorSQLSelect.java:667)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.executeSearchRecord(OCommandExecutorSQLSelect.java:616)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.serialIterator(OCommandExecutorSQLSelect.java:1639)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.fetchFromTarget(OCommandExecutorSQLSelect.java:1554)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.executeSearch(OCommandExecutorSQLSelect.java:511)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.execute(OCommandExecutorSQLSelect.java:474)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.execute(OCommandExecutorSQLDelegate.java:72)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.executeCommand(OAbstractPaginatedStorage.java:2602)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:2548)
	at com.orientechnologies.orient.server.distributed.impl.ODistributedStorage.command(ODistributedStorage.java:269)
	at com.orientechnologies.orient.core.sql.query.OSQLQuery.run(OSQLQuery.java:78)
	at com.orientechnologies.orient.core.sql.query.OSQLAsynchQuery.run(OSQLAsynchQuery.java:74)
	at com.orientechnologies.orient.core.query.OQueryAbstract.execute(OQueryAbstract.java:33)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.query(ODatabaseDocumentTx.java:696)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLUpdate.execute(OCommandExecutorSQLUpdate.java:288)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.execute(OCommandExecutorSQLDelegate.java:72)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.executeCommand(OAbstractPaginatedStorage.java:2602)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:2548)
	at com.orientechnologies.orient.server.distributed.impl.ODistributedStorage.command(ODistributedStorage.java:269)
	at com.orientechnologies.orient.core.command.OCommandRequestTextAbstract.execute(OCommandRequestTextAbstract.java:69)
	at com.carnoot.server.functions.Authentication.heartbeat(Authentication.java:93)
	at com.carnoot.server.WebHandler.handle(WebHandler.java:57)
	at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)
	at sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:83)
	at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:82)
	at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:675)
	at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)
	at sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:647)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.orientechnologies.orient.server.distributed.task.ODistributedRecordLockedException: Record with rid #27:0 is locked by request 0.1
	DB name="Carnoot"
	at com.orientechnologies.orient.server.distributed.impl.ODistributedStorage.acquireRecordLock(ODistributedStorage.java:1260)
	at com.orientechnologies.orient.server.distributed.impl.ODistributedStorage.executeRecordOperationInLock(ODistributedStorage.java:1029)
	at com.orientechnologies.orient.server.distributed.impl.ODistributedStorage.updateRecord(ODistributedStorage.java:798)
	at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:2046)
	at com.orientechnologies.orient.core.tx.OTransactionNoTx.saveRecord(OTransactionNoTx.java:186)
	... 38 more
`

The program receives a HTTP request, and gets the user's ID via their session ID:

`
public class AuthenticatedUser {

    ORID userID;

    public AuthenticatedUser(String sessionID) {
        ODatabase DB = CarnootDB.newDBConnection();
        try {
            if (sessionID != "empty") {

                List<ODocument> queryResult = DB.query(new OSQLSynchQuery("SELECT expand(both('SESSION')) FROM (SELECT FROM SESSIONS WHERE SESSIONID=decode('" + sessionID + "','base64'))"));
                if (queryResult.size() > 0) {
                    userID = queryResult.get(0).getIdentity();
                }
            }
        } finally {
            DB.close();
        }

    }

    public ORID getUser() throws Exception {
        if (userID != null) {
            return userID;
        } else {
            throw new AuthenticationException(AuthenticationException.ErrorType.SESSION_ID_INVALID);
        }
    }

}`

An instance of this class is then passed to the heartbeat function:

`

ODatabase DB = CarnootDB.newDBConnection();
try {
            DB.activateOnCurrentThread();
            DB.command(new OCommandSQL("UPDATE " + user.getUser().toString() + " SET HEARTBEAT=sysdate()")).execute();

        } catch (Exception E) {
            E.printStackTrace();
        } finally {
            DB.close();
        }`

**Comment by lvca** [comment]
Which release are you using?

**Comment by Dacnomanie** [comment]
2.2.4

**Comment by lvca** [comment]
This has been already fixed in 2.2.5-SNAPSHOT (tomorrow the release, but you can already try it by using the 2.2.x branch)

**Comment by lvca** [comment]
Ok, this should be already fixed in 2.2.5-SNAPSHOT.

**Comment by nishantkumar1292** [comment]
Hi @lvca even I am getting this error. Using build **2.2.21**
```
Exception `1B662E37` in storage `db_development`
com.orientechnologies.orient.core.exception.ODatabaseException: Error during saving of record with rid #55:6
        DB name="db_development"
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.saveRecord(OTransactionNoTx.java:203)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:2753)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.save(ODatabaseDocumentTx.java:102)
        at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1805)
        at com.orientechnologies.orient.core.record.impl.ODocument.save(ODocument.java:1796)
        at com.tinkerpop.blueprints.impls.orient.OrientGraphNoTx.addEdgeInternal(OrientGraphNoTx.java:249)
        at com.tinkerpop.blueprints.impls.orient.OrientGraphNoTx.addEdgeInternal(OrientGraphNoTx.java:137)
        at com.tinkerpop.blueprints.impls.orient.OrientVertex.addEdge(OrientVertex.java:741)
        at com.orientechnologies.orient.graph.sql.OCommandExecutorSQLCreateEdge$1.call(OCommandExecutorSQLCreateEdge.java:185)
        at com.orientechnologies.orient.graph.sql.OCommandExecutorSQLCreateEdge$1.call(OCommandExecutorSQLCreateEdge.java:143)
        at com.orientechnologies.orient.graph.sql.OGraphCommandExecutorSQLFactory.runInConfiguredTxMode(OGraphCommandExecutorSQLFactory.java:260)
        at com.orientechnologies.orient.graph.sql.OCommandExecutorSQLCreateEdge.execute(OCommandExecutorSQLCreateEdge.java:143)
        at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.execute(OCommandExecutorSQLDelegate.java:74)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.executeCommand(OAbstractPaginatedStorage.java:3200)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:3138)
        at com.orientechnologies.orient.server.distributed.impl.ODistributedStorage.command(ODistributedStorage.java:260)
        at com.orientechnologies.orient.core.command.OCommandRequestTextAbstract.execute(OCommandRequestTextAbstract.java:69)
        at com.orientechnologies.orient.server.network.protocol.http.command.post.OServerCommandPostBatch.execute(OServerCommandPostBatch.java:171)
        at com.orientechnologies.orient.server.network.protocol.http.ONetworkProtocolHttpAbstract.service(ONetworkProtocolHttpAbstract.java:169)
        at com.orientechnologies.orient.server.network.protocol.http.ONetworkProtocolHttpAbstract.execute(ONetworkProtocolHttpAbstract.java:621)
        at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:77)
Caused by: com.orientechnologies.orient.core.exception.ODatabaseException: Error on saving record #55:6
        DB name="db_development"
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeSaveRecord(ODatabaseDocumentTx.java:2254)
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.saveRecord(OTransactionNoTx.java:181)
        ... 20 more
Caused by: java.lang.NullPointerException
Internal server error:

java.lang.NullPointerException
```
Any updates on how to resolve this.
